// electron-compile/lib/read-only-compiler.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _regenerator = require("babel-runtime/regenerator");

var _regenerator2 = _interopRequireDefault(_regenerator);

var _asyncToGenerator2 = require("babel-runtime/helpers/asyncToGenerator");

var _asyncToGenerator3 = _interopRequireDefault(_asyncToGenerator2);

var _classCallCheck2 = require("babel-runtime/helpers/classCallCheck");

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require("babel-runtime/helpers/createClass");

var _createClass3 = _interopRequireDefault(_createClass2);

var _lodash = require("lodash");

var _lodash2 = _interopRequireDefault(_lodash);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * ReadOnlyCompiler is a compiler which allows the host to inject all of the compiler
 * metadata information so that {@link CompileCache} et al are able to recreate the
 * hash without having two separate code paths.
 */

var ReadOnlyCompiler = function () {
  /**  
   * Creates a ReadOnlyCompiler instance
   *    
   * @private
   */

  function ReadOnlyCompiler(name, compilerVersion, compilerOptions, inputMimeTypes) {
    (0, _classCallCheck3.default)(this, ReadOnlyCompiler);

    _lodash2.default.assign(this, { name: name, compilerVersion: compilerVersion, compilerOptions: compilerOptions, inputMimeTypes: inputMimeTypes });
  }

  (0, _createClass3.default)(ReadOnlyCompiler, [{
    key: "shouldCompileFile",
    value: function () {
      var ref = (0, _asyncToGenerator3.default)(_regenerator2.default.mark(function _callee() {
        return _regenerator2.default.wrap(function _callee$(_context) {
          while (1) {
            switch (_context.prev = _context.next) {
              case 0:
                return _context.abrupt("return", true);

              case 1:
              case "end":
                return _context.stop();
            }
          }
        }, _callee, this);
      }));

      function shouldCompileFile() {
        return ref.apply(this, arguments);
      }

      return shouldCompileFile;
    }()
  }, {
    key: "determineDependentFiles",
    value: function () {
      var ref = (0, _asyncToGenerator3.default)(_regenerator2.default.mark(function _callee2() {
        return _regenerator2.default.wrap(function _callee2$(_context2) {
          while (1) {
            switch (_context2.prev = _context2.next) {
              case 0:
                return _context2.abrupt("return", []);

              case 1:
              case "end":
                return _context2.stop();
            }
          }
        }, _callee2, this);
      }));

      function determineDependentFiles() {
        return ref.apply(this, arguments);
      }

      return determineDependentFiles;
    }()
  }, {
    key: "compile",
    value: function () {
      var ref = (0, _asyncToGenerator3.default)(_regenerator2.default.mark(function _callee3() {
        return _regenerator2.default.wrap(function _callee3$(_context3) {
          while (1) {
            switch (_context3.prev = _context3.next) {
              case 0:
                throw new Error("Read-only compilers can't compile");

              case 1:
              case "end":
                return _context3.stop();
            }
          }
        }, _callee3, this);
      }));

      function compile() {
        return ref.apply(this, arguments);
      }

      return compile;
    }()
  }, {
    key: "shouldCompileFileSync",
    value: function shouldCompileFileSync() {
      return true;
    }
  }, {
    key: "determineDependentFilesSync",
    value: function determineDependentFilesSync() {
      return [];
    }
  }, {
    key: "compileSync",
    value: function compileSync() {
      throw new Error("Read-only compilers can't compile");
    }
  }, {
    key: "getCompilerVersion",
    value: function getCompilerVersion() {
      return this.compilerVersion;
    }
  }]);
  return ReadOnlyCompiler;
}();

exports.default = ReadOnlyCompiler;
